console/src/components/Services/Data/TablePermissions/JSONEditor.tsx

```diff
@@ -0,0 +1,64 @@
+import React, { useState, useEffect } from 'react';
+import AceEditor, { IAnnotation } from 'react-ace';
+import { isJsonString } from '../../../Common/utils/jsUtils';
+import { usePrevious } from '../../../../hooks/usePrevious';
+
+export interface JSONEditorProps {
+  initData: string;
+  onChange: (v: string) => void;
+  data: string;
+}
+
+const JSONEditor: React.FC<JSONEditorProps> = ({
+  initData,
+  onChange,
+  data,
+}) => {
+  const [value, setValue] = useState(initData || data || '');
+  const [annotations, setAnnotations] = useState<IAnnotation[]>([]);
+  const prevData = usePrevious<string>(data);
+
+  useEffect(() => {
+    // if the data prop is changed do nothing
+    if (prevData !== data) return;
+    // when state gets new data, trigger parent callback
+    if (value !== data) onChange(value);
+  }, [value, data, prevData]);
+
+  // check and set error message
+  useEffect(() => {
+    if (isJsonString(value)) {
+      setAnnotations([]);
+    } else {
+      setAnnotations([
+        { row: 0, column: 0, text: 'Invalid JSON', type: 'error' },
+      ]);
+    }
+    return () => {
+      setAnnotations([]);
+    };
+  }, [value]);
+
+  useEffect(() => {
+    // set data to editor only if the prop has a valid json string
+    // setting value from query editor will always have a valid json
+    // any invalid json means, the value is set from this component so no need to set that again
+    if (isJsonString(data)) setValue(data);
+  }, [data]);
+
+  return (
+    <AceEditor
+      mode="json"
+      onChange={setValue}
```

Comment:
It causes console error when typing in the JSON row editor:
<img width="683" alt="Screenshot 2020-06-03 at 17 11 44" src="https://user-images.githubusercontent.com/9019397/83654376-50218f00-a5bd-11ea-824a-5e22136d387b.png">
```suggestion
      onChange={v => setValue(v)}
```

Reply:
This is because of the debounce, will either make a callback function or use an arrow function (I tend to avoid [arrow functions in render ](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md))

Will fix it :)